Allow recovering from tty switching failures

aleixpol opened the pull request:
If we fail to switch tty when showing the greeter, exit saying so then switch back to tty 1 and try launching the Display again.

Fixes #1636

~~Draft because:~~
- ~~we should avoid looping infinitely.~~ works
- ~~we should do something similar for autologin use cases.~~ works
- ~~we should do something similar for user switching use cases.~~ https://github.com/sddm/sddm/issues/1642

aleixpol committed:
Allow recovering from tty switching failures

If we fail to switch tty when showing the greeter, exit saying so then
switch back to tty 1 and try launching the Display again.

Fixes #1636

alebastr commented:
This seems to be working fine in a sane configuration, thanks!

If `ReserveVT=1` is set in `logind.conf` though (as configured in [openSUSE's gdm package](https://build.opensuse.org/package/view_file/openSUSE:Factory/gdm/reserveVT.conf?expand=1)),  the initial boot just quietly fails and subsequent restarts of sddm will either succeed or stuck in an infinite loop.

Vogtinator commented on src/daemon/Display.cpp:
```diff
@@ -149,10 +149,11 @@ namespace SDDM {
+        connect(m_greeter, &Greeter::failed, this, &Display::stop);
+        connect(m_greeter, &Greeter::ttyFailed, this, [this] {
+            VirtualTerminal::jumpToVt(1, true);
+            stop();
```
IMO needs a comment here

Vogtinator commented on src/daemon/Display.cpp:
```diff
@@ -149,10 +149,11 @@ namespace SDDM {
-                });
+        connect(m_greeter, &Greeter::failed, this, &Display::stop);
+        connect(m_greeter, &Greeter::ttyFailed, this, [this] {
+            VirtualTerminal::jumpToVt(1, true);
```
Why tty1? In openSUSE we have a getty running there, this might break it

aleixpol replied:
getty is everywhere in my experience. Do you have an alternative in mind?

aleixpol replied:
Would you prefer having a default tty setting?

arrowd replied:
And on FreeBSD we have getty running on ttys `0-7` and X server starts on `8`.

Vogtinator replied:
On openSUSE it's getty on tty1-tty6 by default (started on the first switch) and X on tty7 (sddm.conf has `MinimumVT=7`).

Because getty is started on demand, tty1-tty6 are actually free until the switch and something like OPENQRY can return them for use by e.g. `kwin_wayland`.

It's not entirely clear to me what the VT number specified here effectively does...

aleixpol replied:
> On openSUSE it's getty on tty1-tty6 by default (started on the first switch) and X on tty7 (sddm.conf has MinimumVT=7).

MinimumVT was dropped a while ago. https://github.com/sddm/sddm/commit/f0b6dec66a6fc8cc6e21da57f81190843b483f76

> Because getty is started on demand, tty1-tty6 are actually free until the switch and something like OPENQRY can return them for use by e.g. kwin_wayland.

Well that kind of is the problem. We sometimes want to have a predictable tty for our session and the greeter. openqry can tell us which one is free but getty takes them over and never releases them so you can easily end up with a session that is all over the place.

> It's not entirely clear to me what the VT number specified here effectively does...

Where you want the graphical sessions to start from. I guess not unlike the MinimumVT setting from back then.

Vogtinator replied:
Yep, we'll definitely need a replacement for that. IMO it's important to have this deterministic and predictable. systemd upstream is unfortunately opposed to that...

aleixpol replied:
> In openSUSE we have a getty running there, this might break it

Unless you were patching out this `Conflicts=getty@tty1.service` in the service file, this statement is not true. If you were patching it, I imagine it was conflicting at times?

In any case, I've added a CMake variable so that you can use start from a different tty.

Vogtinator replied:
> Unless you were patching out this Conflicts=getty@tty1.service in the service file, this statement is not true. If you were patching it, I imagine it was conflicting at times?

Yep (https://build.opensuse.org/package/view_file/openSUSE:Factory/sddm/0001-Systemd-service-unit-Use-tty7-by-default.patch?expand=1)! Though we don't actually use `sddm.service` by default, we have `display-manager.service` which starts the displaymanager (configured in `/etc/sysconfig/displaymanager) itself. That'll probably change soon though.

aleixpol commented:
> If `ReserveVT=1` is set in `logind.conf` though (as configured in [openSUSE's gdm package](https://build.opensuse.org/package/view_file/openSUSE:Factory/gdm/reserveVT.conf?expand=1)), the initial boot just quietly fails and subsequent restarts of sddm will either succeed or stuck in an infinite loop.

Right, it sounds like something that shouldn't be set for sddm's package. Or is it something you'd like to have?

aleixpol committed:
Introduce SDDM_INITIAL_VT as the TTY to reach out to

We were telling sddm to hold onto tty1, this allows us to configure a
different tty.
